### PR TITLE
GitHub-CI: test CPP17

### PR DESCRIPTION
**Pull request by peterrum**
As discussed with @tjhei.

Tempted by https://github.com/dealii/dealii/pull/13497 and #13502.

### CODE REVIEW

**Review by tjhei — APPROVED** [review]
Good idea!

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -68,7 +68,7 @@ jobs:
         mkdir build
         cd build
         cmake -D CMAKE_BUILD_TYPE=Debug \
-              -D DEAL_II_CXX_FLAGS='-Werror' \
+              -D DEAL_II_CXX_FLAGS='-Werror -std=c++17' \
```
Comment (marcfehling) [inline .github/workflows/linux.yml:71]: We could also use the CMAKE functionality. I'm actually undecided what I prefer personally :-)
```suggestion
              -D CMAKE_CXX_STANDARD=17 \
              -D DEAL_II_CXX_FLAGS='-Werror' \
```

**Review by bangerth — COMMENTED** [review]
This just switches one of the testers to C++17. Ideally, or course, we would add a separate CI step that *adds* a C++17 tester, but that increases the wait time we have (or exceeds whatever limit we are allotted). How did you choose this particular tester to switch to C++17? I supposed this leaves enough testers who continue to check C++14.

File: .github/workflows/linux.yml (hunk 1 above)
Comment (peterrum) [reply .github/workflows/linux.yml]: I have changed this!

**Comment by peterrum** [comment]
> This just switches one of the testers to C++17. Ideally, or course, we would add a separate CI step that adds a C++17 tester, but that increases the wait time we have (or exceeds whatever limit we are allotted). How did you choose this particular tester to switch to C++17? I supposed this leaves enough testers who continue to check C++14.

This is how we normally do it. If we see that something is not tested (in the past: python bindings, 64-bit indices), we just modify one test. We still have enough tests with C++14. Adding a new test would be an option, but would mean longer queue times.

**Comment by masterleinad** [comment]
I don't think setting `CMAKE_CXX_STANDARD` works. I still see
```
#        C++ language standard:  C++14
```
in the respective CI run. We should rather do `DEAL_II_CXX_FLAGS=-std=c++17`.

**Comment by marcfehling** [comment]
> I don't think setting `CMAKE_CXX_STANDARD` works. I still see
> 
> ```
> #        C++ language standard:  C++14
> ```
> 
> in the respective CI run. We should rather do `DEAL_II_CXX_FLAGS=-std=c++17`.

Interesting. I think in addition to `CMAKE_CXX_STANDARD` we would need to set `CXX_STANDARD_REQUIRED=ON`, see also [here](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD_REQUIRED.html). Or we can specify the flags for the `gnu` compiler directly...

**Comment by masterleinad** [comment]
> I think in addition to `CMAKE_CXX_STANDARD` we would need to set `CXX_STANDARD_REQUIRED=ON`, see also [here](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD_REQUIRED.html). Or we can specify the flags for the `gnu` compiler directly.

That isn't enough for me. I think I remember that https://github.com/dealii/dealii/pull/10307 ignores the `CMAKE_CXX_STANDARD` completely and only relies on flags we pass and the default behavior of the compiler to decide what kind of language version support we have.

**Comment by peterrum** [comment]
CI says now: `C++ language standard:  C++17`!

**Review by masterleinad — APPROVED** [review]

**Comment by bangerth** [comment]
Nice! Now we just need a tester for C++20 ;-)

**Comment by peterrum** [comment]
> Nice! Now we just need a tester for C++20 ;-)

@bangerth Couldn't we just simply convert the C++17 to a C++20 test? What would we miss?

**Comment by bangerth** [comment]
We probably could, but I don't think we use any C++20 features yet (even with the appropriate `#ifdef`). So I think there is, at least for now, not a lot of use for this.

But that might change in the future, of course. Do we have a tester that already has a compiler that's new enough? I think we'd need GCC 11 at least.

**Pull request merged**